File: DSA/GreedyAlgo/Code/FractionalKnapsack.java
// Fractional Knapsack in Java

import java.util.*;

public class FractionalKnapsack {
    static class Item implements Comparable<Item> {
        int id;
        double weight;
        double value;
        double ratio;
        
        Item(int id, double weight, double value) {
            this.id = id;
            this.weight = weight;
            this.value = value;
            this.ratio = value / weight;
        }
        
        public int compareTo(Item other) {
            return Double.compare(other.ratio, this.ratio);
        }
    }
    
    public static void fractionalKnapsack(Item[] items, double capacity) {
        Arrays.sort(items);
        
        System.out.println("Fractional Knapsack Solution:");
        double totalValue = 0;
        double usedCapacity = 0;
        
        for (Item item : items) {
            if (usedCapacity >= capacity) break;
            
            double take = Math.min(capacity - usedCapacity, item.weight);
            System.out.printf("Item %d: Take %.2f units (Weight: %.2f, Value/Weight: %.2f)%n",
                    item.id, take, item.weight, item.ratio);
            
            totalValue += take * item.ratio;
            usedCapacity += take;
        }
        
        System.out.printf("%nTotal Capacity Used: %.2f / %.2f%n", usedCapacity, capacity);
        System.out.printf("Total Value: %.2f%n", totalValue);
    }
    
    public static void main(String[] args) {
        System.out.println("=== Fractional Knapsack (Java) ===\n");
        
        Item[] items = {
            new Item(1, 10, 60),
            new Item(2, 20, 100),
            new Item(3, 30, 120)
        };
        double capacity = 50;
        
        System.out.println("Items (ID, Weight, Value):");
        for (Item item : items) {
            System.out.printf("Item %d: Weight %.2f, Value %.2f%n", item.id, item.weight, item.value);
        }
        System.out.printf("Knapsack Capacity: %.2f%n%n", capacity);
        
        fractionalKnapsack(items, capacity);
        
        System.out.println("\n=== Greedy Strategy ===");
        System.out.println("1. Calculate value/weight ratio for each item");
        System.out.println("2. Sort items by ratio (descending)");
        System.out.println("3. For each item in sorted order:");
        System.out.println("   - If item fits completely: take it");
        System.out.println("   - Otherwise: take fraction that fits");
        System.out.println("4. Repeat until capacity full\n");
        
        System.out.println("=== Complexity ===");
        System.out.println("Time Complexity:  O(n log n) for sorting");
        System.out.println("Space Complexity: O(1)");
        System.out.println("Optimal: Yes (greedy works for fractional)");
    }
}

File: DSA/GreedyAlgo/Code/FractionalKnapsack.py
# Fractional Knapsack in Python

class Item:
    def __init__(self, id, weight, value):
        self.id = id
        self.weight = weight
        self.value = value
        self.ratio = value / weight

def fractional_knapsack(items, capacity):
    # Sort by value/weight ratio
    items.sort(key=lambda x: x.ratio, reverse=True)
    
    print("Fractional Knapsack Solution:")
    total_value = 0
    used_capacity = 0
    
    for item in items:
        if used_capacity >= capacity:
            break
        
        take = min(capacity - used_capacity, item.weight)
        print(f"Item {item.id}: Take {take:.2f} units (Weight: {item.weight}, Value/Weight: {item.ratio:.2f})")
        
        total_value += take * item.ratio
        used_capacity += take
    
    print(f"\nTotal Capacity Used: {used_capacity:.2f} / {capacity}")
    print(f"Total Value: {total_value:.2f}")

# Test
print("=== Fractional Knapsack (Python) ===\n")

items = [
    Item(1, 10, 60),
    Item(2, 20, 100),
    Item(3, 30, 120)
]
capacity = 50

print("Items (ID, Weight, Value):")
for item in items:
    print(f"Item {item.id}: Weight {item.weight}, Value {item.value}")
print(f"Knapsack Capacity: {capacity}\n")

fractional_knapsack(items, capacity)

print("\n=== Greedy Strategy ===")
print("1. Calculate value/weight ratio for each item")
print("2. Sort items by ratio (descending)")
print("3. For each item in sorted order:")
print("   - If item fits completely: take it")
print("   - Otherwise: take fraction that fits")
print("4. Repeat until capacity full\n")

print("=== Complexity ===")
print("Time Complexity:  O(n log n) for sorting")
print("Space Complexity: O(1)")
print("Optimal: Yes (greedy works for fractional)")
